File: 01 November 2022/C++/Median in a row-wise sorted Matrix.cpp
//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++

class Solution{   
public:
    int median(vector<vector<int>> &matrix, int r, int c){
         int mini = matrix[0][0];
         int maxi = matrix[0][c-1];
         for(int i=1;i<r;i++){
             if(matrix[i][0] < mini){
                 mini = matrix[i][0];
             }
             if(matrix[i][c-1] > maxi)
             {
                 maxi = matrix[i][c-1];
             }
         }
         
         int medPos = (r*c + 1)/2;
         
         while(mini < maxi){
             int mid = (mini + maxi)/2;
             int midPos = 0;
             
             for(int i=0;i<r;i++){
                 midPos += upper_bound(matrix[i].begin(),matrix[i].end(),mid) - matrix[i].begin();
             }
             if(midPos < medPos){
                 mini = mid + 1;
             }
             else{
                 maxi = mid;
             }
         }
         return mini;
    }
};


//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int r, c;
        cin>>r>>c;
        vector<vector<int>> matrix(r, vector<int>(c));
        for(int i = 0; i < r; ++i)
            for(int j = 0;j < c; ++j)
                cin>>matrix[i][j];
        Solution obj;
        cout<<obj.median(matrix, r, c)<<endl;        
    }
    return 0;
}
// } Driver Code Ends

File: 01 November 2022/Python/Median in a row-wise sorted Matrix.py
#User function Template for python3

from bisect import bisect_right
class Solution:
    def median(self, matrix, r, c):
        min=float("inf")
        max=-float("inf")
        for i in range(r):
            if matrix[i][0]<min:
                min=matrix[i][0]
            if matrix[i][c-1]>max:
                max=matrix[i][c-1]
        desired=(r*c+1)//2
        while min<max:
            mid=(max+min)//2
            place=0
            for i in range(r):
                place+=bisect_right(matrix[i],mid)
            if place<desired:
                min=mid+1
            else:
                max=mid
        
        return min  


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    ob = Solution()
    t = int (input ())
    for _ in range (t):
        r,c = map(int,input().strip().split())
        matrix = [[0 for j in range(c)] for i in range(r)]
        for i in range(r):
            t=[int(el) for el in input().split()]
            for j in range(c):
                matrix[i][j]=t[j]
        ans = ob.median(matrix, r, c);
        print(ans)
# } Driver Code Ends